fix: prevent requesting a new trade window with each "trade"

dudantas opened the pull request:
Identified Problems:
• Item Loop Performance: The loop in the ProtocolGame::sendShop function performed key-value store (KV) access operations for each NPC item during every iteration. This was highly inefficient for NPCs with many items, such as "The Lootmonger," which has 1380 items.
• Lack of Database Caching: Currently, our database does not implement caching for query results. This leads to redundant database accesses, particularly noticeable when commands like "trade" are used repeatedly, significantly increasing CPU usage and processing time.

Implemented Solutions:
• Optimization of the Item Loop: Necessary initialization has been moved outside the loop in the ProtocolGame::sendShop method. This reduces overhead by avoiding unnecessary repetitive database accesses.
• Removal of Unnecessary Map: Removed an extraneous map in npc.hpp and simplified the use of the shop player cache.
These changes aim to streamline interactions and reduce the computational load, especially during frequent trade requests with item-rich NPCs like "The Lootmonger." Additional optimizations have been made to enhance overall system performance.

dudantas committed:
fix: do not request a new trade window with each say "trade"

ThiagoNBCoding commented:
@dudantas As you know way more than protocol than I do, I just have a question:

These changes are according to RL Tibia, and they really don't accept further trade requests when a window is already open. But besides that, wouldn't it be possible(and maybe optimal) to add a cache to the protocol?

To further optimize, given a certain NPC, they: have a specific store for each keyword, a specific store for each state of a player quest/world quest/world state or a combination of these things.

Considering that, wouldn't it be possible to, on startup, load and cache into a variable the part of the protocol of each NPC responsible for item prices and the list of items(ontop of weight, etc), thus, removing the need of that **for**(which I assume that for NPCs that have a long list of items, is the costy part). So we'd only need to create the part of the protocol responsible for player data(like ther current gold, for example), and concatenate both before sending.

dudantas commented:
> @dudantas As you know way more than protocol than I do, I just have a question:
> 
> These changes are according to RL Tibia, and they really don't accept further trade requests when a window is already open. But besides that, wouldn't it be possible(and maybe optimal) to add a cache to the protocol?
> 
> To further optimize, given a certain NPC, they: have a specific store for each keyword, a specific store for each state of a player quest/world quest/world state or a combination of these things.
> 
> Considering that, wouldn't it be possible to, on startup, load and cache into a variable the part of the protocol of each NPC responsible for item prices and the list of items(ontop of weight, etc), thus, removing the need of that **for**(which I assume that for NPCs that have a long list of items, is the costy part). So we'd only need to create the part of the protocol responsible for player data(like ther current gold, for example), and concatenate both before sending.

Yes, your logic is correct. Even though my correction really makes sense, because globally it does not resend the trade after it is already open, there is a logic problem in the way that the shop is currently cached at the base, I will correct that.

#edit:
Here we have two problems; first, the loop was running the kv (key-value store) at each iteration, which was very bad if the NPC had many items (in the case of the lootmonger, 1380 items). Second, the database currently does not cache results, and with each query, it accesses the database again, which results in an absurd amount of time and processing being spent in the loop. 

I will correct the first problem; I would also fix the second one, but that requires more time.

dudantas committed:
enhance: moved initialization outside the loop to avoid overhead

Removed an unnecessary map in npc.hpp and simplified the use of the shop player.

Some others enhancements made for optimization.

github-actions[bot] committed:
Code format - (Clang-format)

dudantas committed:
Merge branch 'main' into dudantas/fix-close-shop-window
